### PR TITLE
added group entity and service layers

### PR DESCRIPTION
**Pull request by samidc1992**

### CODE REVIEW

File: entities/Group.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    .populate('admin', 'username -_id');
+
+    return group;
+}
+
+const updateGroupById = async(group_id) => {
+
+    const updateInformation = await Group.updateOne(
+        { _id: group_id },
+        { photo: url }
```
Comment (aymericsecret) [inline entities/Group.js:34]: where does url come from ? It looks undefined
Also if you update the url only, you can rename to `updateGroupPhotoById`

File: services/Group.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
+const GroupEntity = require('../entities/Group');
+const SportEntity = require('../entities/Sport');
+const UserEntity = require('../entities/User');
+
+const Group = require('../models/groups');
+
+const createGroup = async(token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude, status) => {
+
+    const sportData = await SportEntity.findSportBySportName(sport);
```
Comment (aymericsecret) [inline services/Group.js]: Just to be clear, you could rename `sport` by `sportName`

File: services/Group.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    const userData = await UserEntity.findUserByToken(token);
+
+    if(!sportData || !userData) {
+        return;
+    };
+
+    const newGroup = await new Group({
```
Comment (aymericsecret) [inline services/Group.js]: Do you need the await here ? I think class instantiations are not async

File: services/Group.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    if(!sportData) {
+        return;
+    };
+
+    const groupsData = await GroupEntity.findGroupBySportAndOrLocation(sport, sportData._id, latitude, longitude);
+
+    return groupsData;
+};
+
+const getGroupInformation = async(group_id) => {
```
Comment (aymericsecret) [inline services/Group.js:56]: Better to use camelCase in the code (same in `getGroupMembers`) !

File: routes/groups.js [hunk 5]
```diff
@@ -6,121 +6,87 @@ const cloudinary = require('cloudinary').v2;
...
-router.post('/create', (req, res) => {
+router.post('/create', async(req, res) => {
 
-    let { token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude } = req.body;
+    let { token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude, status } = req.body;
 
-
-    if (!token && !photo && !name && !sport && !maxMembers && !genders && !levels && !ageMin && !ageMax && !description && !label && !latitude && !longitude) {
-        res.json({ result: false, message: 'Missing or empty fields.' });
+    if (!token || !photo || !name || !sport || !maxMembers || !genders || !levels || !ageMin || !ageMax || !description || !label || !latitude || !longitude) {
```
Comment (aymericsecret) [inline routes/groups.js:15]: `status` is not mandatory ?

File: entities/User.js [hunk 6]
```diff
@@ -11,70 +11,75 @@ const findUserByUsername = async (username) => {
...
-              status: "Approved",
-            }
+const removeUserFromGroup = async (token, group_id) => {
+  const user = await User.updateOne(
+      { token },
+      {
+        $pull: {
+          registrations: {
+            group: group_id,
+            status: "Approved",
```
Comment (aymericsecret) [inline entities/User.js:77]: If not approved, you don't want to remove the user ?

File: routes/groups.js [hunk 7]
```diff
@@ -6,121 +6,87 @@ const cloudinary = require('cloudinary').v2;
...
-                    User.updateOne(
-                        { token: token },
-                        { $push: { registrations: { group: newEntry._id, status: "Approved" } } }
-                    ).then(() => {
-                        res.json({ result: true, message: 'New group created successfully.', data: newEntry });
-                    })
-                });
-            });
-    })
+    const newGroupData = await GroupService.createGroup(token, photo, name, sport, maxMembers, genders, levels, ageMin, ageMax, description, label, latitude, longitude, status);
```
Comment (aymericsecret) [inline routes/groups.js:20]: One thing you could do next is add error handling. For instance, what will happen here if `GroupService.createGroup` throws an error (you can test is by adding at the beginning of the `createGroup` function `throw new Error('oups')` and see what happens.

Then you should add 
```
try { 
  // your code the can throw error, like `await GroupService.createGroup()`
} catch (err) { 
  // error handling here) 
}
```

Then you can think about where to handle errors. Should you let entity layer and service layer throw errors, and catch them in your routes ? Or should you catch entity layer errors in your service layer ? Or catch mongoose errors in your entity error, and handle it there ?

File: entities/Group.js (hunk 1 above)
Comment (samidc1992) [reply entities/Group.js:34]: true, I did not use the updateGroupById entity yet but I will add url and change the name as suggested.

File: entities/User.js (hunk 6 above)
Comment (samidc1992) [reply entities/User.js:77]: this entity allows for the user that is already in a group (with status approved) to leave the group if/when he chooses to. In the future, if we have an admin approval workflow, users will be able to send a request to join a group and then we can either create a separated entity for users to be able to cancel the request or look into updating this one. I would say it's better to separate them as use cases are slightly different, do you have a recommendation?

File: routes/groups.js (hunk 5 above)
Comment (samidc1992) [reply routes/groups.js:15]: yes, it is not mandatory, as of now we are setting the status approved by default as we did not yet create an approval workflow.

File: services/Group.js (hunk 3 above)
Comment (samidc1992) [reply services/Group.js]: I am still not sure whether class instantiations are async or not, but I will dig further. For now, I will keep it (to be safe) but remove once I confirm they are not needed.

File: services/Group.js (hunk 4 above)
Comment (samidc1992) [reply services/Group.js:56]: Where do you see camelCase missing?

File: services/Group.js (hunk 2 above)
Comment (samidc1992) [reply services/Group.js]: Done

File: routes/groups.js (hunk 7 above)
Comment (samidc1992) [reply routes/groups.js:20]: Thanks, I would say the earlier we catch an error the better, and throwing errors at each layer will definetily help us identify where the error is coming from. So I would catch the errors for every single layer: 1) catch mongoose errors at entity level, 2) catch business logic errors at service level and 3) finally catch errors from the controller at the route level. Is that the correct approach?

File: services/Group.js (hunk 4 above)
Comment (aymericsecret) [reply services/Group.js:56]: snakecase: group_id

File: routes/groups.js (hunk 5 above)
Comment (aymericsecret) [reply routes/groups.js:15]: ok but it's not in the list :p

File: entities/User.js (hunk 6 above)
Comment (aymericsecret) [reply entities/User.js:77]: ok ! when you implement it, you can re evaluate if it makes sense to make a generic function or not

File: services/Group.js (hunk 3 above)
Comment (samidc1992) [reply services/Group.js]: Removed await from class instantiations.

File: services/Group.js (hunk 4 above)
Comment (samidc1992) [reply services/Group.js:56]: Done!

File: routes/groups.js (hunk 5 above)
Comment (samidc1992) [reply routes/groups.js:15]: it's set by default in the users model (see backend > models > users.js)

**Pull request merged**